File: src/mlogix/logix/Token.java
package mlogix.logix;

import mlogix.compiler.struct.Span;

public class Token {
	public final TokenType type;
	public final Span span;
	public final Object literal;

	public Token(TokenType type, Span span) {
		this.type = type;
		this.span = span;
		this.literal = null;
	}
	
	public Token(TokenType type, Span span, Object literal) {
		this.type = type;
		this.span = span;
		this.literal = literal;
	}
	
	public String toString() {
		return String.format("Token{%s,%s,%s}", type.name(), span.toString(), literal);
	}

	public String toSimpleString() {
		return String.format("Token(%s,%s)", type.name(), literal);
	}

}

File: src/mlogix/compiler/core/token/Token.kt
package mlogix.compiler.core.token

import mlogix.compiler.core.span.Span
import mlogix.compiler.core.span.Spanned

data class Token(
    val span: Span,
    val type: TokenType,
    val literal: Any? = null
) : Spanned {

    override fun toString(): String =
        "Token{${type.name},${span},$literal}"

    fun toSimpleString(): String =
        "Token(${type.name},$literal)"

    override fun span(): Span {
        return span
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Token

        if (type != other.type) return false
        if (literal != other.literal) return false

        return true
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + literal.hashCode()
        return result
    }
}